src/hotspot/share/utilities/align.hpp

```diff
@@ -72,7 +72,9 @@ constexpr T align_down(T size, A alignment) {
 
 template<typename T, typename A, ENABLE_IF(std::is_integral<T>::value)>
 constexpr T align_up(T size, A alignment) {
-  T adjusted = checked_cast<T>(size + alignment_mask(alignment));
+  T mask = checked_cast<T>(alignment_mask(alignment));
+  assert(size <= std::numeric_limits<T>::max() - mask, "overflow");
```

Comment:
I don't really understand this assertion. `align_up((uint32_t)0fffff_ffff, 16) == 0`, because `uint32_t` is an unsigned type:

_An unsigned integer type has the same width N as the corresponding signed integer type. The range of representable values for the unsigned type is 0 to 2 N − 1 (inclusive); arithmeticfor the unsigned type is performed modulo 2**N_.
[Note 2 : Unsigned arithmetic does not overflow. Overflow for signed arithmetic yields undefined behavior ]

Reply:
The JBS issue is using "overflow" in the sense of "high bits of the mathematical result are discarded".
Fixed-width unsigned arithmetic can certainly overflow in that sense.

Reply:
Perhaps, although "overflow" has a precisze definition in C++, but I would be extremely surprised if I were looking for the address of the end of the page at 0fffff_ffff and was informed of an overflow. IMO, _there is no overflow_ in that case, and the correct answer must be 0.

Reply:
There's no "perhaps" about the intended meaning in the JBS issue. I wrote that
issue; I remember what I meant. :) I suppose I could have been more precise.

So I disagree. I think align_up has an implied post-condition that the result
is not less than the value being aligned. That's certainly how it's used, in
every occurrance I've looked at. (I admit I didn't look at all ~450 uses
though.)

Reply:
BTW @theRealAph , regarding your example of the address of the the end of the page at the end of memory,
pointer arithmetic is not modular, and instead overflow is UB.

Reply:
> BTW @theRealAph , regarding your example of the address of the the end of the page at the end of memory, pointer arithmetic is not modular, and instead overflow is UB.

I wasn't talking about pointer arithmetic, as was explicit in my reply. I was talking about unsigned integer types.

Reply:
That wasn't obvious, because of "the address of the end of the page at".  I took that to mean pointers were involved.